beps/0002-dynamic-frontend-plugins/README.md

```diff
@@ -154,12 +154,386 @@ Each plugin is required to provide manifest file (metadata) in predefined format
 
-## Design Details
+### CLI
+
+The following sections contains proposal of new CLI scripts to support the dynamic frontend plugins.
+
+#### Static and Dynamic Plugin Build Differences
+
+The major difference between static and dynamic plugins lies in their dependencies and how they are "installed" into a Backstage instance and whether they are bundled into final JS output during the build.
+
+While static plugins can define their dependencies using the dependencies field inside `package.json`, dynamic (module federation-based) plugins can't.
+
+A dynamic plugin requires all dependencies to be present in the build output. Module sharing is then used to optimize the "output size" at runtime. If a shared package is not found in the browser scope, or if the version range does not match the requirement, it falls back and loads additional JS assets from the "private" build and adds them to the scope.
+
+Webpack is also not the best tool for building traditional NPM packages. Dependencies can be externalized, but Webpack brings additional code into the bundles. Using Webpack external dependencies also conflicts with module federation shared modules. Externals will not be included even as a shared module fallback. As a result, such shared module can't be used because the dependency will not be defined in the init container.
+
+If the vision is to still support "static" plugins installed via NPM, it makes sense to keep the static and dynamic plugin builds separate and build two different bundles. Both can still be published as a single NPM package (not recommended), but to achieve optimal performance for both, there cannot be a shared output.
+
+#### Dynamic export CLI
+
+```sh
+backstage-cli plugin export-dynamic
+```
+
+Similar to the current build command, instead of producing static plugin output, it builds a dynamic plugin using Webpack and module federation plugins.
+
+**Arguments**
+
+- _path_ (optional)
+  - path to the plugin `package.json`
+  - defaults to `cwd`
+- _export-name_
+  - Export name from where the plugin root is accessible
+  - defaults to `pluginEntry`
+- _cwd_
+  - current working directory of the script
+  - defaults to current `cwd` value
+
+**Shared modules**
+
+> NOTE: Details of module sharing are described in the [Module sharing](#module-sharing) section.
+
+Short summary of dependency sharing:
+
+- Predefined set of singleton packages
+- Generated granular config of modules for MUI dependencies
+- Share all other dependencies listed in package.json
+
+To allow additional customization of module sharing, extra configuration will be required. A similar approach to the current catalog-info.yaml config file can be used. There is also the option of embedding the metadata in the `package.json`. However, the configuration can become very large and clutter the package.json.
```

Comment:
Do you have an example in mind where a significant amount of config would be required?

As a side note, the existing `catalog-info.yaml` files don't really have anything to do with the Backstage build tooling, it's used for tracking in the catalog and shouldn't be used as an excuse for adding more YAML 😁

Reply:
> the existing catalog-info.yaml files don't really have anything to do with the Backstage build tooling,

I was not aware of that. 

> Do you have an example in mind where a significant amount of config would be required?

I think whenever some component library (like MUI) is present in a plugin. I also think at some point plugin builders might want to exclude specific dependencies from sharing. I personally don't like adding anything to `package.json` that is not related to npm (I've seen to many jest configs taking over the whole file). But that's my personal preference. If you prefer using the file, I don't have any objective argument why not.

Reply:
I'd say typically you would not change the defaults at all. The vast majority should be covered by the default settings. I'll change the section and propose some package.json entries. Do you already have some backstage build-related fields reserved in `package.json`?